library/Focus.hs

```diff
@@ -49,21 +51,48 @@ instance MonadTrans (Focus element) where
 data Change a
   = -- | Produce no changes
     Leave
-  | -- | Delete it
-    Remove
   | -- | Set its value to the provided one
     Set a
-  deriving (Functor, Eq, Ord, Show)
+  | -- | Delete it
+    Remove
+  deriving (Functor, Foldable, Eq, Ord, Show)
 
 instance Semigroup (Change a) where
   (<>) l r =
     case r of
       Leave -> l
       _ -> r
+  stimes = stimesIdempotentMonoid
 
 instance Monoid (Change a) where
   mempty = Leave
 
+
+-- | Determine if the @Change@ is a @Set@
+{-# INLINE isSet #-}
+isSet :: Change a -> Bool
+isSet c = isTrue# (dataToTag# c)
+
+{-# INLINE maybeToChange #-}
+maybeToChange :: Maybe a -> Change a
+maybeToChange = unsafeCoerce
+
+{-# INLINE changeToMaybe #-}
+changeToMaybe :: Change a -> Maybe a
+changeToMaybe = \case
+  Set a -> Just a
+  _     -> Nothing
```

Comment:
*Sight-* no coerce because we have an extra constructor.

```haskell
changeToMaybe :: Change a -> Maybe a
changeToMaybe =
  modifyTag# (and# 1##)

modifyTag# ∷ (Word# -> Word#) -> a -> b
modifyTag# f (unsafeCoerce#->c) = unsafeCoerce# do
    and# c ptr_mask `or#` f (and# c tag_mask `minusWord#` 1##) `plusWord#` 1##
  where
    tag_bits = 3# -- 2 if 32-bit
    tag_mask = (shiftL# 1## tag_bits) `minusWord#` 1##
    ptr_mask = not# tag_mask
```